integration/app_integration_test.go

```diff
@@ -0,0 +1,430 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			statusCode, message, err := pack.makeRequest(tt.inCookie)
+			require.NoError(t, err)
+			require.Equal(t, tt.outStatusCode, statusCode)
+			require.Contains(t, message, tt.outMessage)
+		})
+	}
+}
+
+// pack contains identity as well as initialized Teleport clusters and instances.
+type pack struct {
+	username string
+	password string
+
+	rootCluster        *TeleInstance
+	rootAppServer      *service.TeleportProcess
+	rootAppName        string
+	rootAppPublicAddr  string
+	rootAppClusterName string
+	rootMessage        string
+
+	leafCluster        *TeleInstance
+	leafAppServer      *service.TeleportProcess
+	leafAppName        string
+	leafAppPublicAddr  string
+	leafAppClusterName string
+	leafMessage        string
+}
+
+// setup configures all clusters and servers needed for a test.
+func setup(t *testing.T) *pack {
+	utils.InitLoggerForTests(testing.Verbose())
+
+	tr := utils.NewTracer(utils.ThisFunction()).Start()
+	defer tr.Stop()
+
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
+
+	SetTestTimeouts(time.Millisecond * time.Duration(500))
+
+	p := &pack{
+		rootAppName:        "app-01",
+		rootAppPublicAddr:  "app-01.example.com",
+		rootAppClusterName: "example.com",
+		rootMessage:        uuid.New(),
+
+		leafAppName:        "app-02",
+		leafAppPublicAddr:  "app-02.example.com",
+		leafAppClusterName: "leaf.example.com",
+		leafMessage:        uuid.New(),
+	}
+
+	// Start HTTP server that will be acting like an proxied application.
+	rootServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, p.rootMessage)
+	}))
+	t.Cleanup(func() {
+		rootServer.Close()
+	})
+	leafServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, p.leafMessage)
+	}))
+	t.Cleanup(func() {
+		leafServer.Close()
+	})
+
+	var err error
+
+	privateKey, publicKey, err := testauthority.New().GenerateKeyPair("")
+	require.NoError(t, err)
+
+	// Find AllocatePortsNum free listening ports to use.
+	startNumber := utils.PortStartingNumber + (AllocatePortsNum * 2) + 1
+	ports, err := utils.GetFreeTCPPorts(AllocatePortsNum, startNumber+1)
+	require.NoError(t, err)
+
+	// Create a new Teleport instance with passed in configuration.
+	p.rootCluster = NewInstance(InstanceConfig{
+		ClusterName: "example.com",
+		HostID:      uuid.New(),
+		NodeName:    Host,
+		Ports:       ports.PopIntSlice(5),
+		Priv:        privateKey,
+		Pub:         publicKey,
+	})
+
+	// Create a new Teleport instance with passed in configuration.
+	p.leafCluster = NewInstance(InstanceConfig{
+		ClusterName: "leaf.example.com",
+		HostID:      uuid.New(),
+		NodeName:    Host,
+		Ports:       ports.PopIntSlice(5),
+		Priv:        privateKey,
+		Pub:         publicKey,
+	})
+
+	rcConf := service.MakeDefaultConfig()
+	rcConf.DataDir, err = ioutil.TempDir("", "cluster-"+p.rootCluster.Secrets.SiteName)
```

Comment:
Temporary directories created here and on lines 195, 222 and 246 below don't seem to be cleaned up after the test.